core/controller/src/main/scala/whisk/core/loadBalancer/InvokerSupervision.scala

```diff
@@ -150,14 +238,99 @@ class InvokerActor extends FSM[InvokerState, InvokerInfo] {
         case Event(StateTimeout, _)   => goto(Offline)
     }
 
+    /**
+     * Handle the completion of an Activation in every state.
+     */
+    whenUnhandled {
+        case Event(cm: InvocationFinishedMessage, info) => handleCompletionMessage(cm.successful, info.buffer)
+    }
+
+    /** Logging on Transition change */
+    onTransition {
+        case _ -> Offline   => transid.mark(this, LoggingMarkers.LOADBALANCER_INVOKER_OFFLINE, s"$name is offline", akka.event.Logging.WarningLevel)
+        case _ -> UnHealthy => transid.mark(this, LoggingMarkers.LOADBALANCER_INVOKER_UNHEALTHY, s"$name is unhealthy", akka.event.Logging.WarningLevel)
+        case _ -> Healthy   => logging.info(this, s"$name is healthy")
+    }
+
+    /** Scheduler to send testActivations, if the invoker is UnHealthy */
     onTransition {
-        case Healthy -> Offline => logging.warn(this, s"$name is down")
-        case Offline -> Healthy => logging.info(this, s"$name is online")
+        case _ -> UnHealthy => {
+            invokeTestAction()
+            setTimer(InvokerActor.timerName, Tick, 1.minute, true)
+        }
+        case UnHealthy -> _ => cancelTimer(InvokerActor.timerName)
     }
 
     initialize()
+
+    /**
+     * Handling for active acks. This method saves the result (successful or unsuccessful)
+     * into an RingBuffer and checks, if the InvokerActor has to be changed to UnHealthy.
+     *
+     * @param wasActivationSuccessful: result of Activation
+     * @param buffer to be used
+     */
+    private def handleCompletionMessage(wasActivationSuccessful: Boolean, buffer: RingBuffer[Boolean]) = {
+        buffer.add(wasActivationSuccessful)
+
+        // If the current state is UnHealthy, this means, that the activeAck is from an testaction.
+        // If this is successful it seems like the Invoker is Healthy again. So we execute immediately
+        // a new testaction to remove the errors out of the RingBuffer as fast as possible.
+        if (wasActivationSuccessful && stateName == UnHealthy) {
+            invokeTestAction()
```

Comment:
i'm wondering if it's necessary to have this `if` clause (279-281) since the `else` clause below will check if the number of activations is above a threshold and the actor will remain in the unhealthy state until the threshold is exceeded.

the comment suggests that the this is proactively sending another test action (one more more) - vs letting the timer (1 minute intervals) fire a next one... is this violating the actor contract (namely, should there be a slow pole (1 min) and then a faster reactive one? In essence you can end up sending some additional test actions here that pile up with the ones from the timer.

This may be acceptable - but if that's deliberate, we should expand the comments further.